src/execution/operator/join/physical_index_join.cpp

```diff
@@ -0,0 +1,214 @@
+	vector<vector<row_t>> rhs_rows;
+	ExpressionExecutor probe_executor;
+	IndexLock lock;
+};
+
+PhysicalIndexJoin::PhysicalIndexJoin(LogicalOperator &op, unique_ptr<PhysicalOperator> left,
+                                     unique_ptr<PhysicalOperator> right, vector<JoinCondition> cond, JoinType join_type,
+                                     const vector<idx_t> &left_projection_map, vector<idx_t> right_projection_map,
+                                     vector<column_t> column_ids_, Index *index, bool lhs_first)
+    : PhysicalOperator(PhysicalOperatorType::INDEX_JOIN, move(op.types)), left_projection_map(left_projection_map),
+      right_projection_map(move(right_projection_map)), index(index), conditions(move(cond)), join_type(join_type),
+      lhs_first(lhs_first) {
+	column_ids = move(column_ids_);
+	children.push_back(move(left));
+	children.push_back(move(right));
+	for (auto &condition : conditions) {
+		condition_types.push_back(condition.left->return_type);
+	}
+	//! Only add to fetch_ids columns that are not indexed
+	for (auto &index_id : index->column_ids) {
+		index_ids.insert(index_id);
+	}
+	for (size_t column_id = 0; column_id < column_ids.size(); column_id++) {
+		auto it = index_ids.find(column_ids[column_id]);
+		if (it == index_ids.end()) {
+			fetch_ids.push_back(column_ids[column_id]);
+			fetch_types.push_back(children[1]->types[column_id]);
+		}
+	}
+}
+
+void PhysicalIndexJoin::Output(ExecutionContext &context, DataChunk &chunk, PhysicalOperatorState *state_) {
+	auto &transaction = Transaction::GetTransaction(context.client);
+	auto &phy_tbl_scan = (PhysicalTableScan &)*children[1];
+	auto &bind_tbl = (TableScanBindData &)*phy_tbl_scan.bind_data;
+	auto tbl = bind_tbl.table->storage.get();
+	DataChunk rhs_chunk;
+	idx_t rhs_column_idx = 0;
+	SelectionVector sel;
+	sel.Initialize(STANDARD_VECTOR_SIZE);
+	size_t output_sel_idx{};
+	vector<row_t> fetch_rows;
+	auto state = reinterpret_cast<PhysicalIndexJoinOperatorState *>(state_);
+	while (output_sel_idx < STANDARD_VECTOR_SIZE && state->lhs_idx < state->child_chunk.size()) {
+		if (state->rhs_idx < state->result_sizes[state->lhs_idx]) {
+			sel.set_index(output_sel_idx++, state->lhs_idx);
+			if (!fetch_types.empty()) {
+				//! We need to collect the rows we want to fetch
+				fetch_rows.push_back(state->rhs_rows[state->lhs_idx][state->rhs_idx]);
+			}
+			state->rhs_idx++;
+		} else {
+			//! We are done with the matches from this LHS Key
+			state->rhs_idx = 0;
+			state->lhs_idx++;
+		}
+	}
+	//! Now we fetch the RHS data
+	if (!fetch_types.empty()) {
+		rhs_chunk.Initialize(fetch_types);
+		ColumnFetchState fetch_state;
+		Vector row_ids;
+		row_ids.type = LOGICAL_ROW_TYPE;
+		FlatVector::SetData(row_ids, (data_ptr_t)&fetch_rows[0]);
+		tbl->Fetch(transaction, rhs_chunk, fetch_ids, row_ids, output_sel_idx, fetch_state);
+	}
+
+	//! Now we actually produce our result chunk
+	if (!lhs_first) {
+		for (idx_t i = 0; i < right_projection_map.size(); i++) {
```

Comment:
As far as I can tell all that changes based on `lhs_first` is the index into the final chunk, is that correct? If so, rather than duplicating code perhaps you can do:
```sql
idx_t left_offset = lhs_first ? 0 : right_projection_map.size();
idx_t right_offset = lhs_first ? left_projection_map.size() : 0;
...
chunk.data[left_offset + i].Reference(...)
...
chunk.data[right_offset + i].Reference(...)
```